docs/reference/fields.txt

```diff
@@ -1160,41 +1160,26 @@ you can include the short versions of the modules.
 Field Names with Dots/Periods (``.``) and Dollar Signs (``$``)
 ==============================================================
 
-MongoDB query language (MQL) generally uses the dot/period character (``.``)
-to separate field names in a field path that traverses embedded documents, and
-the dollar character (``$``) for operators. It is possible to utilize field
-names containing the dot and dollar characters, however it is recommended to
-avoid these characters for ease of querying.
-
-MongoDB 5.0 adds `special operators
+At present, neither Mongoid nor the Mongo Ruby Driver support inserting
+or querying against fields whose name contains the dot/period (``.``)
+character or begins with the dollar sign (``$``) character.
```

Comment:
This is not correct. Inserting dots for example:

```
irb(main):018:0> client=Mongo::Client.new(['localhost:15020'])
=> #<Mongo::Client:0x11220 cluster=#<Cluster topology=ReplicaSetWithPrimary[localhost:15020,localhost:1...
irb(main):019:0> client['foo'].delete_many
=> #<Mongo::Operation::Delete::Result:0x11480 documents=[{"n"=>1, "electionId"=>BSON::ObjectId('7fffffff0000000000000001'), "opTime"=>{"ts"=>#<BSON::Timestamp:0x00007f79da417a98 @seconds=1648068293, @increment=1>, "t"=>1}, "ok"=>1.0, "$clusterTime"=>{"clusterTime"=>#<BSON::Timestamp:0x00007f79da417930 @seconds=1648068293, @increment=1>, "signature"=>{"hash"=><BSON::Binary:0x11500 type=generic data=0x0000000000000000...>, "keyId"=>0}}, "operationTime"=>#<BSON::Timestamp:0x00007f79da417728 @seconds=1648068293, @increment=1>}]>
irb(main):020:0> client['foo'].insert_one('_id' => 4)
=> #<Mongo::Operation::Insert::Result:0x11880 documents=[{"n"=>1, "electionId"=>BSON::ObjectId('7fffffff0000000000000001'), "opTime"=>{"ts"=>#<BSON::Timestamp:0x00007f79da41e230 @seconds=1648068294, @increment=1>, "t"=>1}, "ok"=>1.0, "$clusterTime"=>{"clusterTime"=>#<BSON::Timestamp:0x00007f79da41e0c8 @seconds=1648068294, @increment=1>, "signature"=>{"hash"=><BSON::Binary:0x11900 type=generic data=0x0000000000000000...>, "keyId"=>0}}, "operationTime"=>#<BSON::Timestamp:0x00007f79da41dec0 @seconds=1648068294, @increment=1>}]>
irb(main):021:0> client['foo'].update_one({'_id'=>4},'a.b.c'=>5)
=> #<Mongo::Operation::Update::Result:0x12560 documents=[{"n"=>1, "electionId"=>BSON::ObjectId('7fffffff0000000000000001'), "opTime"=>{"ts"=>#<BSON::Timestamp:0x00007f79dc1fce78 @seconds=1648068295, @increment=1>, "t"=>1}, "nModified"=>1, "ok"=>1.0, "$clusterTime"=>{"clusterTime"=>#<BSON::Timestamp:0x00007f79dc1fc680 @seconds=1648068295, @increment=1>, "signature"=>{"hash"=><BSON::Binary:0x12580 type=generic data=0x0000000000000000...>, "keyId"=>0}}, "operationTime"=>#<BSON::Timestamp:0x00007f79dc1fc0e0 @seconds=1648068295, @increment=1>}]>
irb(main):022:0> client['foo'].find.to_a
=> [{"_id"=>4, "a.b.c"=>5}]
```

Reply:
@p-mongo is this just the driver or does inserting work in Mongoid as well?

Reply:
```
class Band
  include Mongoid::Document
  
  field 'foo.bar', type: Integer
  field '$hello', type: Integer
end
```

Mongoid with current bson-rbuy prohibiting dots/dollars:

```
irb(main):029:0> b = Band.create!
=> #<Band _id: 623c8090a15d5d34df623ad2, foo.bar: nil, $hello: nil>
irb(main):030:0> b.collection
=> #<Mongo::Collection:0x13140 namespace=mongoid.bands>
irb(main):031:0> b.collection.update_one({_id:b.id}, 'foo.bar'=>42)
=> #<Mongo::Operation::Update::Result:0x13620 documents=[{"n"=>1, "nModified"=>1, "ok"=>1.0}]>
irb(main):032:0> Band.last
=> #<Band _id: 623c8090a15d5d34df623ad2, foo.bar: 42, $hello: nil>
```

With the following patch applied to bson-ruby that removes the restriction for hash serialization:

```
diff --git a/ext/bson/write.c b/ext/bson/write.c
index 4268fa6..dc080a0 100644
--- a/ext/bson/write.c
+++ b/ext/bson/write.c
@@ -333,7 +333,7 @@ void pvt_put_bson_key(byte_buffer_t *b, VALUE string, VALUE validating_keys){
   
   if (RTEST(validating_keys)) {
     if (length > 0 && (c_str[0] == '$' || memchr(c_str, '.', length))) {
-      rb_exc_raise(rb_funcall(rb_bson_illegal_key, rb_intern("new"), 1, string));
+      //rb_exc_raise(rb_funcall(rb_bson_illegal_key, rb_intern("new"), 1, string));
     }
   }
```

Mongoid:

```
irb(main):011:0> b = Band.create!('foo.bar'=>42)
=> #<Band _id: 623c8286a15d5d393a236252, foo.bar: 42, $hello: nil>
irb(main):012:0> Band.last
=> #<Band _id: 623c8286a15d5d393a236252, foo.bar: 42, $hello: nil>
irb(main):013:0> Band.last.attributes
=> {"_id"=>BSON::ObjectId('623c8286a15d5d393a236252'), "foo.bar"=>42}
```

Reply:
Is the same also true of fields beginning with dollars? i.e. Ruby driver allows them?

Reply:
(Assuming the answer is yes, I've updated the language accordingly.